Introduce a script to generate TestWebKitAPI entitlements

aestes opened the pull request:
#### 2177a497f6efeb5d73a550f771900ef35b7d86ea
<pre>
Introduce a script to generate TestWebKitAPI entitlements
<a href="https://bugs.webkit.org/show_bug.cgi?id=291618">https://bugs.webkit.org/show_bug.cgi?id=291618</a>
<a href="https://rdar.apple.com/problem/149363565">rdar://problem/149363565</a>

Reviewed by Elliott Williams.

To make it easier to add unique entitlements to TestWebKitAPI.app without duplicating the
entitlements from TestWebKitAPI, introduced a process-entitlements.sh script to generate
entitlements for TestWebKitAPI and TestWebKitAPI.app. Ensured that non-base entitlements are added
to a __TEXT section in simulator binaries and to the code signature otherwise.

Verified that TestWebKitAPI and TestWebKitAPI.app have the same entitlements before and after this
change.

* Tools/TestWebKitAPI/Configurations/TestWebKitAPI-iOS.entitlements: Removed.
* Tools/TestWebKitAPI/Configurations/TestWebKitAPI-macOS-internal.entitlements: Removed.
* Tools/TestWebKitAPI/Configurations/TestWebKitAPI-macOS.entitlements: Removed.
* Tools/TestWebKitAPI/Configurations/TestWebKitAPIBase.xcconfig:
* Tools/TestWebKitAPI/Scripts/process-entitlements.sh: Added.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/294089@main">https://commits.webkit.org/294089@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3e3098ffd5d6190cf2c669eb6b0ca596d93a58b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/100797 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/20449 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/10748 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/105934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/51386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/102838 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/20758 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/28923 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/105934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/51386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/103804 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/20758 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/138/builds/10748 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/105934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/20758 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/10748 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/50761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/20758 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/10748 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/108289 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/27915 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/28923 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/108289 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/28278 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/138/builds/10748 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/108289 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21697 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/10748 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/21919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/27850 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/27661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/30979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/29219 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/TestWebKitAPI/Scripts/process-entitlements.sh:
```diff
@@ -0,0 +1,93 @@
+rm -f "${WK_PROCESSED_XCENT_FILE}"
+rm -f "${WK_PROCESSED_BASE_XCENT_FILE}"
+plistbuddy Clear dict
+plistbuddy_base Clear dict
```
nit: Do you really need to rm -f here? Seems like `PlistBuddy Clear ...` will reset the plist for you.

emw-apple commented on Tools/TestWebKitAPI/Configurations/TestWebKitAPIBase.xcconfig:
```diff
@@ -144,12 +144,17 @@ OTHER_LDFLAGS_PLATFORM_cocoatouch = -framework WebCore -framework UIKit -framewo
+OTHER_LDFLAGS_ENTITLEMENTS[sdk=*simulator] = -Xlinker -sectcreate -Xlinker __TEXT -Xlinker __entitlements -Xlinker $(WK_PROCESSED_XCENT_FILE);
+
+WK_PROCESSED_BASE_XCENT_FILE = ;
+WK_PROCESSED_BASE_XCENT_FILE[sdk=*simulator] = $(TEMP_FILE_DIR)/$(FULL_PRODUCT_NAME).xcent;
```
naming nit: I wonder if it would be clearer to give this a simulator-specific name, since its purpose is to hold entitlements that apply to the code signature of simulated binaries only. how about `WK_SIMULATED_XCENT_FILE`?

emw-apple commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -6538,6 +6546,7 @@
 			isa = PBXNativeTarget;
 			buildConfigurationList = 1DEB927408733DD40010E9CD /* Build configuration list for PBXNativeTarget "TestWebKitAPI" */;
 			buildPhases = (
+				A1508C552DAF1F8000C3E213 /* Process Entitlements */,
```
Out of curiosity, do we _need_ to run the script this early, before compilation? I would have expected it to run post-compiling, pre-linking, i.e. between the Sources and Frameworks build phases.

No big deal either way.

aestes replied:
I'm not sure I totally understand why, but a cycle is introduced when this phase comes after Compile Sources.

aestes committed:
Introduce a script to generate TestWebKitAPI entitlements
https://bugs.webkit.org/show_bug.cgi?id=291618
rdar://problem/149363565

Reviewed by Elliott Williams.

To make it easier to add unique entitlements to TestWebKitAPI.app without duplicating the
entitlements from TestWebKitAPI, introduced a process-entitlements.sh script to generate
entitlements for TestWebKitAPI and TestWebKitAPI.app. Ensured that non-base entitlements are added
to a __TEXT section in simulator binaries and to the code signature otherwise.

Verified that TestWebKitAPI and TestWebKitAPI.app have the same entitlements before and after this
change.

* Tools/TestWebKitAPI/Configurations/TestWebKitAPI-iOS.entitlements: Removed.
* Tools/TestWebKitAPI/Configurations/TestWebKitAPI-macOS-internal.entitlements: Removed.
* Tools/TestWebKitAPI/Configurations/TestWebKitAPI-macOS.entitlements: Removed.
* Tools/TestWebKitAPI/Configurations/TestWebKitAPIBase.xcconfig:
* Tools/TestWebKitAPI/Scripts/process-entitlements.sh: Added.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/294089@main
